salt/orch/update.sls

```diff
@@ -265,25 +286,46 @@ early-services-setup:
     - kwarg:
         destructive: True
     - require:
-      - {{ master_id }}-start-services
-
-{% endfor %}
+      - all-masters-post-start-services
+{%- endfor %}
 
-# Perform migrations after all masters have been updated
-all-masters-post-start-services:
+# NOTE: Remove me for 4.0
+#
+# On 2.0 -> 3.0 we are updating the way kubelets auth against the apiservers.
+# At this point in time all masters have been updated, and all workers are (or
+# will) be in `NotReady` state. This means that any operation that we perform
+# that go through the apiserver down to the kubelets won't work (e.g. draining
+# nodes).
+#
+# To fix this problem we'll apply the haproxy sls to all worker nodes, so they
+# can rejoin the cluster and we can operate on them normally.
+all-workers-2.0-pre-clean-shutdown:
   salt.state:
-    - tgt: '{{ is_master_tgt }}'
+    - tgt: '( {{ is_updateable_worker_tgt }} ) and G@osrelease:2.0'
     - tgt_type: compound
-    - batch: 3
     - sls:
-      - cni.update-post-start-services
-      - kubelet.update-post-start-services
+        - etc-hosts
+        - migrations.2-3.haproxy
     - require:
-      - early-services-setup
+        - all-masters-post-start-services
 {%- for master_id in masters.keys() %}
       - {{ master_id }}-reboot-needed-grain
 {%- endfor %}
 
+# Sanity check. If an operator manually rebooted a machine when it had the 3.0
+# snapshot ready, we are already in 3.0 but with an unapplied haproxy config.
+# Apply the main haproxy sls to 3.0 workers (if any).
+all-workers-3.0-pre-clean-shutdown:
+  salt.state:
+    - tgt: '( {{ is_updateable_worker_tgt }} ) and G@osrelease:3.0'
+    - tgt_type: compound
+    - sls:
+        - etc-hosts
+        - haproxy
```

Comment:
Is this `tgt` matching anything? This `tgt` means _"all the 3.0 machines were we have installed updated but they still need to be rebooted"_, and I think the `tx_update_reboot_needed` is cleaned up on reboot by some `service`, right?. So I think it will not really match anything...

Anyway, I think in this kind of scenarios it is better to just highstate instead of running some key (and maybe conflicting) SLSs.

Reply:
`tx_update_reboot_needed` is only cleaned by a script in the admin node, in the rest of the machines it's still there until we remove it with salt (that is, when the upgrade on that machine is known to have succeeded).

This is a special case: imagine a machine had downloaded the 3.0 new snapshot and didn't apply it yet. The operator for whatever reason reboots the machine, it will apply the new snapshot -- it's a corner case, weird to happen, but it was free to protect from. We just apply `haproxy` on all worker nodes (the old version on 2.0 and the new on 3.0 [if any]).

Reply:
Ok, this looks like another undocumented, magic, corner case... (facepalm)